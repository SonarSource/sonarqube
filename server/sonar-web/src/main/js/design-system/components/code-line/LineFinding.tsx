/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import styled from '@emotion/styled';
import { forwardRef, Ref } from 'react';
import tw from 'twin.macro';
import { themeBorder, themeColor, themeContrast, themeShadow } from '../../helpers/theme';
import { BareButton } from '../../sonar-aligned/components/buttons';

interface Props {
  as?: React.ElementType;
  className?: string;
  getFixButton?: React.ReactNode;
  issueKey: string;
  message: React.ReactNode;
  onIssueSelect?: (issueKey: string) => void;
  selected?: boolean;
}

function LineFindingFunc(
  { as, getFixButton, message, issueKey, selected = true, className, onIssueSelect }: Props,
  ref: Ref<HTMLButtonElement>,
) {
  return selected ? (
    <LineFindingStyled
      as="div"
      className={className}
      data-issue={issueKey}
      ref={ref}
      selected={selected}
    >
      {onIssueSelect ? (
        <BareButton
          onClick={() => {
            onIssueSelect(issueKey);
          }}
        >
          {message}
        </BareButton>
      ) : (
        message
      )}
      {getFixButton}
    </LineFindingStyled>
  ) : (
    <LineFindingStyled
      as={as}
      className={className}
      data-issue={issueKey}
      onClick={() => {
        if (onIssueSelect) {
          onIssueSelect(issueKey);
        }
      }}
      ref={ref}
      selected={selected}
    >
      {message}
    </LineFindingStyled>
  );
}

export const LineFinding = forwardRef<HTMLElement, Props>(LineFindingFunc);

const LineFindingStyled = styled(BareButton)<{ selected: boolean }>`
  ${tw`sw-flex sw-gap-2 sw-items-center`}
  ${tw`sw-my-3 sw-mx-1`}
  ${tw`sw-rounded-1`}
  ${tw`sw-px-3`}
  ${tw`sw-w-full`}
  ${tw`sw-box-border`}
  ${(props) => (props.selected ? tw`sw-py-3` : tw`sw-py-2`)};
  ${(props) => (props.selected ? tw`sw-typo-lg-semibold` : tw`sw-typo-default`)};
  ${(props) => (props.selected ? tw`sw-cursor-default` : tw`sw-cursor-pointer`)};

  border: ${(props) =>
    props.selected
      ? themeBorder('default', 'issueBoxSelectedBorder')
      : themeBorder('default', 'issueBoxBorder')};
  color: ${themeContrast('pageBlock')};
  word-break: break-word;
  background-color: ${themeColor('pageBlock')};

  :hover {
    box-shadow: ${themeShadow('sm')};
  }
`;
