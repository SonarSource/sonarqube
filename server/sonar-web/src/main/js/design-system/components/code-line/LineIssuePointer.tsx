/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import styled from '@emotion/styled';
import { createRef, RefObject, useEffect, useState } from 'react';
import tw from 'twin.macro';
import { themeColor } from '../../helpers/theme';

const POINTER_HANDLE_HEIGHT_OFFSET = 2;
const POINTER_HANDLE_MAX_HEIGHT = 10;

interface Props {
  issueFindingRef?: RefObject<HTMLDivElement>;
}

export function LineIssuePointer({ issueFindingRef }: Props) {
  const [distance, setDistance] = useState(0);
  const pointerRef = createRef<HTMLDivElement>();

  useEffect(() => {
    if (pointerRef.current && issueFindingRef?.current) {
      setDistance(
        issueFindingRef.current.getBoundingClientRect().top -
          pointerRef.current.getBoundingClientRect().bottom,
      );
    }
  }, [pointerRef, issueFindingRef]);

  // Keep only the pointer head as reference for future calculations
  if (distance < POINTER_HANDLE_HEIGHT_OFFSET || distance > POINTER_HANDLE_MAX_HEIGHT) {
    return <EmptyIssuePointer data-testid="empty-issue-pointer" ref={pointerRef} />;
  }

  return <IssuePointer data-testid="issue-pointer" distance={distance} ref={pointerRef} />;
}

const EmptyIssuePointer = styled.div`
  bottom: -0.325rem;
  left: 50%;
  width: 0.3125rem;
  height: 0.3125rem;
  border: 0.125rem solid transparent;
  transform: translate(-50%, 0);

  ${tw`sw-block sw-absolute sw-rounded-2`};
`;

const IssuePointer = styled(EmptyIssuePointer)<{ distance: number }>`
  background-color: ${themeColor('codeLineIssueSquiggle')};
  border-color: ${themeColor('codeLineIssuePointerBorder')};

  &::after {
    position: absolute;
    top: 5px;
    left: 2px;
    display: block;
    width: 1px;
    height: ${({ distance }) => `${distance + POINTER_HANDLE_HEIGHT_OFFSET}px`};
    background-color: ${themeColor('codeLineIssueSquiggle')};
    content: '';
  }
`;
