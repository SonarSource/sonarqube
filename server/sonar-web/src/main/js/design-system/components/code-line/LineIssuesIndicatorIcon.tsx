/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import styled from '@emotion/styled';
import { memo } from 'react';
import tw from 'twin.macro';
import { DotFillIcon } from '../icons';

interface Props {
  issuesCount: number;
}

function LineIssueIndicatorIconFunc({ issuesCount }: Readonly<Props>) {
  return (
    <>
      <DotFillIcon />
      {issuesCount > 1 && <IssueIndicatorCounter>{issuesCount}</IssueIndicatorCounter>}
    </>
  );
}

export const LineIssuesIndicatorIcon = memo(LineIssueIndicatorIconFunc);

const IssueIndicatorCounter = styled.span`
  font-size: 0.5rem;
  line-height: 0.5rem;

  ${tw`sw-ml-1/2`}
  ${tw`sw-align-top`}
`;
