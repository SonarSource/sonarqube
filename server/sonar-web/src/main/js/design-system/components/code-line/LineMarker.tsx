/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import styled from '@emotion/styled';
import classNames from 'classnames';
import { forwardRef, Ref, useCallback, useRef } from 'react';
import tw from 'twin.macro';
import { themeColor, themeContrast } from '../../helpers/theme';
import { LocationMarker } from '../LocationMarker';

interface Props {
  hideLocationIndex?: boolean;
  index: number;
  leading?: boolean;
  message?: string;
  onLocationSelect?: (index: number) => void;
  selected: boolean;
}

function LineMarkerFunc(
  { hideLocationIndex, index, leading, message, onLocationSelect, selected }: Props,
  ref: Ref<HTMLElement>,
) {
  const element = useRef<HTMLDivElement | null>(null);
  const elementMessage = useRef<HTMLDivElement | null>(null);

  const handleClick = useCallback(() => {
    onLocationSelect?.(index);
  }, [index, onLocationSelect]);

  return (
    <Wrapper className={classNames({ leading })} ref={element}>
      <LocationMarker
        onClick={handleClick}
        ref={ref as React.RefObject<HTMLDivElement>}
        selected={selected}
        text={hideLocationIndex ? undefined : index + 1}
      />
      {message && <Message ref={elementMessage}>{message}</Message>}
    </Wrapper>
  );
}

const Message = styled.div`
  ${tw`sw-absolute`}
  ${tw`sw-typo-default`}
  ${tw`sw-rounded-1/2`}
  ${tw`sw-px-1`}
  ${tw`sw-left-0`}

  z-index: 1;
  bottom: calc(100% + 0.25rem);
  width: max-content;
  max-width: var(--max-width);
  color: ${themeContrast('codeLineIssueMessageTooltip')};
  background-color: ${themeColor('codeLineIssueMessageTooltip')};
  visibility: hidden;

  &.message-right {
    ${tw`sw-left-auto`}
    ${tw`sw-right-0`}
  }
`;

const Wrapper = styled.div`
  ${tw`sw-relative`}
  ${tw`sw-inline-block`}
  ${tw`sw-align-top`}

  &:not(:first-of-type) {
    ${tw`sw-ml-1`}
  }

  &.leading {
    margin-left: calc(var(--width) - 0.25rem);
  }

  &:hover ${Message} {
    visibility: visible;
  }
`;

export const LineMarker = forwardRef<HTMLElement, Props>(LineMarkerFunc);
