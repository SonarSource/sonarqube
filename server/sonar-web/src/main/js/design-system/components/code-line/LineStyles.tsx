/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import styled from '@emotion/styled';
import tw from 'twin.macro';
import { themeBorder, themeColor, themeContrast } from '../../helpers/theme';
import { BareButton } from '../../sonar-aligned';

export const SCMHighlight = styled.h6`
  color: ${themeColor('tooltipHighlight')};

  ${tw`sw-typo-semibold`};
  ${tw`sw-text-right`};
  ${tw`sw-min-w-[6rem]`};
  ${tw`sw-mr-4`};
  ${tw`sw-my-1`};
`;

export const LineSCMStyledDiv = styled.div`
  outline: none;

  ${tw`sw-pr-2`}
  ${tw`sw-truncate`}
  ${tw`sw-whitespace-nowrap`}
  ${tw`sw-cursor-pointer`}
  ${tw`sw-w-full sw-h-full`}

  &:hover {
    color: var(--echoes-color-text-bold);
  }
`;

export const DuplicationHighlight = styled.h6`
  color: ${themeColor('tooltipHighlight')};

  ${tw`sw-mb-2 sw-font-semibold`};
`;

export const LineStyled = styled.tr`
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: var(--columns);
  align-items: center;

  ${tw`sw-code`}
`;
LineStyled.displayName = 'LineStyled';

export const LineMeta = styled.td`
  color: var(--echoes-color-text-subdued);
  background-color: var(--line-background);
  outline: none;

  ${tw`sw-w-full sw-h-full`}
  ${tw`sw-box-border`}
  ${tw`sw-select-none`}

  ${LineStyled}:hover & {
    background-color: ${themeColor('codeLineHover')};
  }
`;

export const LineCodePreFormatted = styled.pre`
  position: relative;
  white-space: pre-wrap;
  overflow-wrap: anywhere;
  tab-size: 4;
`;

export const LineCodeLayer = styled.div`
  grid-row: 1;
  grid-column: 1;
`;

export const LineCodeLayers = styled.td`
  position: relative;
  display: grid;
  height: 100%;
  background-color: var(--line-background);
  border-left: ${themeBorder('default', 'codeLineBorder')};

  ${LineStyled}:hover & {
    background-color: ${themeColor('codeLineHover')};
  }
`;

export const NewCodeUnderline = styled(LineCodeLayer)`
  background-color: ${themeColor('codeLineNewCodeUnderline')};
`;

export const CoveredUnderline = styled(LineCodeLayer)`
  background-color: ${themeColor('codeLineCoveredUnderline')};
`;

export const UncoveredUnderline = styled(LineCodeLayer)`
  background-color: ${themeColor('codeLineUncoveredUnderline')};
`;

export const UnderlineLabels = styled.div<{ transparentBackground?: boolean }>`
  ${tw`sw-absolute`}
  ${tw`sw-flex sw-gap-1`}
  ${tw`sw-px-1`}
  ${tw`sw-right-0`}


  height: 1.125rem;
  margin-top: -1.125rem;
  background-color: ${({ transparentBackground, theme }) =>
    themeColor(transparentBackground ? 'transparent' : 'codeLine')({ theme })};
`;

export const UnderlineLabel = styled.span`
  ${tw`sw-rounded-t-1`}
  ${tw`sw-px-1`}
`;

export const NewCodeUnderlineLabel = styled(UnderlineLabel)`
  color: ${themeContrast('codeLineNewCodeUnderline')};
  background-color: ${themeColor('codeLineNewCodeUnderline')};
`;

export const CoveredUnderlineLabel = styled(UnderlineLabel)`
  color: ${themeContrast('codeLineCoveredUnderline')};
  background-color: ${themeColor('codeLineCoveredUnderline')};
`;

export const UncoveredUnderlineLabel = styled(UnderlineLabel)`
  color: ${themeContrast('codeLineUncoveredUnderline')};
  background-color: ${themeColor('codeLineUncoveredUnderline')};
`;

export const LineCodeEllipsisStyled = styled(BareButton)`
  ${tw`sw-flex sw-items-center sw-gap-2`}
  ${tw`sw-px-2 sw-py-1`}
  ${tw`sw-code`}
  ${tw`sw-w-full`}
  ${tw`sw-box-border`}

  border-top: ${themeBorder('default', 'codeLineBorder')};
  border-bottom: ${themeBorder('default', 'codeLineBorder')};

  color: var(--echoes-color-text-subdued);
  background-color: ${themeColor('codeLineEllipsis')};

  &:hover {
    background-color: ${themeColor('codeLineEllipsisHover')};
  }
`;
