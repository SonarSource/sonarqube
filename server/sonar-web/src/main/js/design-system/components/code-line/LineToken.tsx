/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import styled from '@emotion/styled';
import classNames from 'classnames';
import { ReactNode, RefObject } from 'react';
import tw from 'twin.macro';
import { themeColor } from '../../helpers/theme';
import { LineIssuePointer } from './LineIssuePointer';

export interface TokenModifiers {
  isHighlighted?: boolean;
  isLocation?: boolean;
  isSelected?: boolean;
  isUnderlined?: boolean;
}

interface Props extends TokenModifiers {
  children: ReactNode;
  className?: string;
  hasMarker?: boolean;
  issueFindingRef?: RefObject<HTMLDivElement>;
}

export function LineToken(props: Props) {
  const { children, className, hasMarker, issueFindingRef, ...modifiers } = props;

  return (
    <TokenStyled
      className={classNames(className, {
        'issue-underline': modifiers.isUnderlined,
        'issue-location': modifiers.isLocation,
        highlighted: modifiers.isHighlighted,
        selected: modifiers.isSelected,
        'has-marker': hasMarker,
      })}
    >
      <>{children}</>
      {modifiers.isUnderlined && <LineIssuePointer issueFindingRef={issueFindingRef} />}
    </TokenStyled>
  );
}

const TokenStyled = styled.span`
  display: inline-block;

  &.sym {
    ${tw`sw-cursor-pointer`}
  }

  &.sym.highlighted {
    background-color: ${themeColor('codeLineLocationHighlighted')};
    transition: background-color 0.3s;
  }

  &.issue-underline {
    position: relative;
    z-index: 1;
    text-decoration: underline ${themeColor('codeLineIssueSquiggle')};
    text-decoration: underline ${themeColor('codeLineIssueSquiggle')} wavy;
    text-decoration-thickness: 2px;
    text-decoration-skip-ink: none;
  }

  &.issue-location {
    line-height: 1.125rem;
    background-color: ${themeColor('codeLineIssueLocation')};
    transition: background-color 0.3s ease;
  }

  &.issue-location.selected {
    background-color: ${themeColor('codeLineIssueLocationSelected')};
  }

  &.issue-location.has-marker {
    ${tw`sw-pl-1`}
  }
`;
