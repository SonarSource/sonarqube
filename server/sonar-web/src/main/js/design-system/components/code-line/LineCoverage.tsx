/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { useTheme } from '@emotion/react';
import styled from '@emotion/styled';
import React, { memo } from 'react';
import tw from 'twin.macro';
import { PopupPlacement } from '../../helpers/positioning';
import { themeColor } from '../../helpers/theme';
import { Tooltip } from '../Tooltip';
import { LineMeta } from './LineStyles';

interface Props {
  coverageStatus?: 'uncovered' | 'partially-covered' | 'covered';
  lineNumber: number;
  scrollToUncoveredLine?: boolean;
  status: string | undefined;
}

function LineCoverageFunc({ lineNumber, coverageStatus, status, scrollToUncoveredLine }: Props) {
  const coverageMarker = React.useRef<HTMLTableCellElement>(null);
  React.useEffect(() => {
    if (scrollToUncoveredLine && coverageMarker.current) {
      coverageMarker.current.scrollIntoView({
        behavior: 'smooth',
        block: 'center',
        inline: 'center',
      });
    }
  }, [scrollToUncoveredLine, coverageMarker]);

  if (!coverageStatus) {
    return <LineMeta data-line-number={lineNumber} />;
  }

  return (
    <Tooltip content={status} placement={PopupPlacement.Right}>
      <LineMeta data-line-number={lineNumber} ref={coverageMarker}>
        {coverageStatus === 'covered' && <CoveredBlock aria-label={status} />}
        {coverageStatus === 'uncovered' && <UncoveredBlock aria-label={status} />}
        {coverageStatus === 'partially-covered' && <PartiallyCoveredBlock aria-label={status} />}
      </LineMeta>
    </Tooltip>
  );
}

export const LineCoverage = memo(LineCoverageFunc);

const CoverageBlock = styled.div`
  ${tw`sw-w-1 sw-h-full`}
  ${tw`sw-ml-1/2`}

  &, & svg {
    outline: none;
  }
`;

const CoveredBlock = styled(CoverageBlock)`
  background-color: ${themeColor('codeLineCovered')};
`;

const UncoveredBlock = styled(CoverageBlock)`
  background-color: ${themeColor('codeLineUncovered')};
`;

function PartiallyCoveredBlock(htmlProps: React.HTMLAttributes<HTMLDivElement>) {
  const theme = useTheme();
  return (
    <CoverageBlock {...htmlProps}>
      <svg fill="none" viewBox="0 0 4 18" xmlns="http://www.w3.org/2000/svg">
        <rect fill={themeColor('codeLinePartiallyCoveredA')({ theme })} height="18" width="4" />
        <path
          clipRule="evenodd"
          d="M0 0L4 3V6L0 3V0ZM0 6L4 9V12L0 9V6ZM4 15L0 12V15L4 18V15Z"
          fill={themeColor('codeLinePartiallyCoveredB')({ theme })}
          fillRule="evenodd"
        />
      </svg>
    </CoverageBlock>
  );
}
