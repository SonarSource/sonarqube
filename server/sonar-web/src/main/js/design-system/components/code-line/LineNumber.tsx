/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import styled from '@emotion/styled';
import { memo, useState } from 'react';
import tw from 'twin.macro';
import { PopupPlacement, PopupZLevel } from '../../helpers/positioning';
import { DropdownToggler } from '../DropdownToggler';
import { LineMeta } from './LineStyles';

interface Props {
  ariaLabel: string;
  displayOptions: boolean;
  firstLineNumber: number;
  lineNumber: number;
  popup: React.ReactNode;
}

const FILE_TOP_THRESHOLD = 10;

function LineNumberFunc({ firstLineNumber, lineNumber, popup, displayOptions, ariaLabel }: Props) {
  const [isOpen, setIsOpen] = useState<boolean>(false);

  const hasLineNumber = Boolean(lineNumber);
  const isFileTop = lineNumber - FILE_TOP_THRESHOLD < firstLineNumber;

  if (!hasLineNumber) {
    return <LineMeta className="sw-pl-2" />;
  }

  return (
    <LineMeta className="sw-pl-2" data-line-number={lineNumber}>
      {displayOptions ? (
        <DropdownToggler
          aria-labelledby={`line-number-trigger-${lineNumber}`}
          id={`line-number-dropdown-${lineNumber}`}
          onRequestClose={() => {
            setIsOpen(false);
          }}
          open={isOpen}
          overlay={popup}
          placement={isFileTop ? PopupPlacement.Bottom : PopupPlacement.Top}
          zLevel={PopupZLevel.Global}
        >
          <LineNumberStyled
            aria-controls={`line-number-dropdown-${lineNumber}`}
            aria-expanded={isOpen}
            aria-haspopup="menu"
            aria-label={ariaLabel}
            id={`line-number-trigger-${lineNumber}`}
            onClick={() => {
              setIsOpen(true);
            }}
            role="button"
            tabIndex={0}
          >
            {lineNumber}
          </LineNumberStyled>
        </DropdownToggler>
      ) : (
        lineNumber
      )}
    </LineMeta>
  );
}

export const LineNumber = memo(LineNumberFunc);

const LineNumberStyled = styled.div`
  outline: none;

  ${tw`sw-pr-2`}
  ${tw`sw-cursor-pointer`}

  &:hover {
    color: var(--echoes-color-text-bold);
  }
`;
