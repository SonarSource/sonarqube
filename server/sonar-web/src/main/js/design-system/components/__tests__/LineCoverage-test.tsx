/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { render } from '../../helpers/testUtils';
import { FCProps } from '../../types/misc';
import { LineCoverage } from '../code-line/LineCoverage';

it('should render correctly when covered', () => {
  expect(setupWithProps().container).toMatchSnapshot();
});

it('should render correctly when uncovered', () => {
  expect(
    setupWithProps({ lineNumber: 16, coverageStatus: 'uncovered' }).container,
  ).toMatchSnapshot();
});

it('should render correctly when partially covered without conditions', () => {
  expect(
    setupWithProps({
      lineNumber: 16,
      coverageStatus: 'partially-covered',
    }).container,
  ).toMatchSnapshot();
});

it('should render correctly when partially covered with 5/10 conditions', () => {
  expect(
    setupWithProps({
      lineNumber: 16,
      coverageStatus: 'partially-covered',
    }).container,
  ).toMatchSnapshot();
});

it('should render correctly when no data', () => {
  expect(setupWithProps({ lineNumber: 16, coverageStatus: undefined }).container).toMatchSnapshot();
});

function setupWithProps(props: Partial<FCProps<typeof LineCoverage>> = {}) {
  return render(<LineCoverage coverageStatus="covered" lineNumber={16} status="OK" {...props} />, {
    container: document.createElement('tr'),
  });
}
