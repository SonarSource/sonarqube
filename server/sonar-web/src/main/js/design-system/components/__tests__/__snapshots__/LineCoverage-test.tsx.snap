// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly when covered 1`] = `
.emotion-0 {
  color: var(--echoes-color-text-subdued);
  background-color: var(--line-background);
  outline: none;
  height: 100%;
  width: 100%;
  box-sizing: border-box;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.e19zkvle13:hover .emotion-0 {
  background-color: rgb(239,242,249);
}

.emotion-2 {
  height: 100%;
  width: 0.25rem;
  margin-left: 0.125rem;
  background-color: rgb(166,244,197);
}

.emotion-2,
.emotion-2 svg {
  outline: none;
}

<tr>
  <td
    aria-describedby="tooltip-1"
    class="emotion-0 emotion-1"
    data-line-number="16"
  >
    <div
      aria-label="OK"
      class="emotion-2 emotion-3"
    />
  </td>
</tr>
`;

exports[`should render correctly when no data 1`] = `
.emotion-0 {
  color: var(--echoes-color-text-subdued);
  background-color: var(--line-background);
  outline: none;
  height: 100%;
  width: 100%;
  box-sizing: border-box;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.e19zkvle13:hover .emotion-0 {
  background-color: rgb(239,242,249);
}

<tr>
  <td
    class="emotion-0 emotion-1"
    data-line-number="16"
  />
</tr>
`;

exports[`should render correctly when partially covered with 5/10 conditions 1`] = `
.emotion-0 {
  color: var(--echoes-color-text-subdued);
  background-color: var(--line-background);
  outline: none;
  height: 100%;
  width: 100%;
  box-sizing: border-box;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.e19zkvle13:hover .emotion-0 {
  background-color: rgb(239,242,249);
}

.emotion-2 {
  height: 100%;
  width: 0.25rem;
  margin-left: 0.125rem;
}

.emotion-2,
.emotion-2 svg {
  outline: none;
}

<tr>
  <td
    aria-describedby="tooltip-4"
    class="emotion-0 emotion-1"
    data-line-number="16"
  >
    <div
      aria-label="OK"
      class="emotion-2 emotion-3"
    >
      <svg
        fill="none"
        viewBox="0 0 4 18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="rgb(217,45,32)"
          height="18"
          width="4"
        />
        <path
          clip-rule="evenodd"
          d="M0 0L4 3V6L0 3V0ZM0 6L4 9V12L0 9V6ZM4 15L0 12V15L4 18V15Z"
          fill="rgb(255,255,255)"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </td>
</tr>
`;

exports[`should render correctly when partially covered without conditions 1`] = `
.emotion-0 {
  color: var(--echoes-color-text-subdued);
  background-color: var(--line-background);
  outline: none;
  height: 100%;
  width: 100%;
  box-sizing: border-box;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.e19zkvle13:hover .emotion-0 {
  background-color: rgb(239,242,249);
}

.emotion-2 {
  height: 100%;
  width: 0.25rem;
  margin-left: 0.125rem;
}

.emotion-2,
.emotion-2 svg {
  outline: none;
}

<tr>
  <td
    aria-describedby="tooltip-3"
    class="emotion-0 emotion-1"
    data-line-number="16"
  >
    <div
      aria-label="OK"
      class="emotion-2 emotion-3"
    >
      <svg
        fill="none"
        viewBox="0 0 4 18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="rgb(217,45,32)"
          height="18"
          width="4"
        />
        <path
          clip-rule="evenodd"
          d="M0 0L4 3V6L0 3V0ZM0 6L4 9V12L0 9V6ZM4 15L0 12V15L4 18V15Z"
          fill="rgb(255,255,255)"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </td>
</tr>
`;

exports[`should render correctly when uncovered 1`] = `
.emotion-0 {
  color: var(--echoes-color-text-subdued);
  background-color: var(--line-background);
  outline: none;
  height: 100%;
  width: 100%;
  box-sizing: border-box;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.e19zkvle13:hover .emotion-0 {
  background-color: rgb(239,242,249);
}

.emotion-2 {
  height: 100%;
  width: 0.25rem;
  margin-left: 0.125rem;
  background-color: rgb(217,45,32);
}

.emotion-2,
.emotion-2 svg {
  outline: none;
}

<tr>
  <td
    aria-describedby="tooltip-2"
    class="emotion-0 emotion-1"
    data-line-number="16"
  >
    <div
      aria-label="OK"
      class="emotion-2 emotion-3"
    />
  </td>
</tr>
`;
